### ISSUE TITLE
BF 3.1.5 - PIDs not showing

### ISSUE THREAD

**Issue by sawmrad** [betaflight/betaflight-tx-lua-scripts#16]
Hello!
Since upgrading to BF 3.5.1, with Telemetry enabled, SmartPort on UART 2 i cannot get pids to show any longer.

XRacer F303 v3.1
XSR RX
Taranis X9D Plus

Have upgraded to latest v0.2 and latest release of OpenTX (RC10 N362).

Any help is appreciated for troubleshooting!

**Comment by ronerone** [comment]
I don't have this issue, same BF, same script, only have to arm to get values got from telemetry.
It correctly displays if changed while it's armed, but i don't know if it really applies without dearming/rearming. i will give it a try this weekend.

**Comment by raphaelcoeffic** [comment]
@ronerone it seems you are sharing SmartPort with MSP, right?
@sawmrad does the SmartPort telemetry work otherwise? Can you see all the values correctly displayed on the telemetry config screen?

**Comment by ronerone** [comment]
@raphaelcoeffic yes it seems so, even if i'm not sure of what it exaclty means, you already pointed this out in another discussion ;)

**Comment by raphaelcoeffic** [comment]
@ronerone that basically means that the same UART is used for SmartPort as well as to connect for example with the Betaflight configurator. On that UART, MSP is enabled at the same time as SmartPort. When disarmed, MSP is enabled and SmartPort is disabled. When armed, MSP is disabled and SmartPort is enabled. Which board are you on?

**Comment by ronerone** [comment]
you're right, this is the behavior. my main telemetry wire is on smartport 1 and my tramp hv T port is on uart 2 if i remember well.
i have a SP F3 racing board.

**Comment by raphaelcoeffic** [comment]
Ok, UART1 is used by the USB interface, and thus should be used for anything else. The new software serial code will be delivered as part of BF 3.1.6, so that you will be able to move your SmartPort connection to UART2, and the TrampHV to a software serial connection (probably motor output 8).

**Comment by ronerone** [comment]
sorry, i have corrected in the same time, i think i'm not on uart1 but on I/O2; is this the same issue?
regarding BF3.1.6, that sounds good! but that mean i'll have to reopen the quad... which will be tricky regarding my dodgy building!

**Comment by sawmrad** [comment]
@raphaelcoeffic - i dont see anything at all. no pids, nothing. just my regular vbatt telemetry works & even that cuts in and out a lot... i don't know why. i've rewired the whole board, which seemed to fix it but it still drops and gives me 'sensor lost' on my taranis.

im using a different uart for smartport.

what do you think?

**Comment by raphaelcoeffic** [comment]
@sawmrad Well, stable SmartPort is a requirement for the scripts to work properly. So I suggest that you start with fixing this. Once this works properly, it should just work. On your board, having SmartPort wired to UART2 TX should work without issues. I guess you have SBUS on UART3 RX. UART1 is not an option, as it is shared with the USB port.

**Comment by si618** [comment]
You can run SmartPort via soft serial if you run out of UARTs, [This approach is working](https://www.rcgroups.com/forums/showthread.php?2793214-Modify-Betaflight-PID-VTX-Settings-w-FrSky-TARANIS-LUA-Script-Menu/page4#post36770475) for a few of us.

**Comment by raphaelcoeffic** [comment]
@si618 BF 3.1.6 is going to change all that (very close to release). I already configured some friends' quads with TrampHV or TBS Unify on softserial (new code able to handle single wire half-duplex), thus keeping SmartPort on hardware UART, which is the best combination given the fact that the bit rate is rather high for soft serial. You will notice that when using 8k/8k, or even 8k/4k.

**Comment by sawmrad** [comment]
@raphaelcoeffic you are correct. i run it off UART2 TX and SBUS is on UART3 RX. UART1 is not an option correct.

I dont know what else i can do to get it to work...
You'd think if regular vbatt works then the lua would work also.

Not sure what else to do here.

**Comment by trojanGoat** [comment]
I just ran into this problem.  This is an issue with the firmware on the XSR.  You need to flash it with a newer one that support PID on telemetry.  

https://www.frsky-rc.com/wp-content/uploads/2017/07/Firmware/XSR/XSR-170619.zip

Tutorial from RG:
https://www.youtube.com/watch?v=od-Qrd4Lulo&lc=z23dcdhh3pbstjblmacdp4311neopuhtk5kenkaftu5w03c010c.1501725754951117

I had the same issue, problem is fixed after flashing new firmware.

**Comment by clcorbin** [comment]
I'm currently fighting this battle.  I have flashed the XSR to the firmware above, Set halfduplex off, running latest BF 3.2, etc.  Telemetery APPEARS to be working as it went from four items (all from the received itself to around 13 items, including things like AccX, AccY, AccZ, Hdg, etc.  They DO appear to update slowly however.

The LUA script appears to be working fine, but all the values listed are simply "---".  My FC is a Kakute F4 AIO and the S. Port wire from the XSR is soldered to the dedicated Smart Port pad on the FC.  And it appears to be working...   Except for those silly PIDs.

**Comment by Dutchy-79** [comment]
Is there a sensor called F000 

If that's not there on the transmitter then you don't have the right version. Might go down one version to try.

**Comment by clcorbin** [comment]
I have seen that one.  I just checked and it is number 3 on the list with a value of 0 and an ID of 27.  And just for fun, I rechecked the LUA script and there are still on "---" for all values.  

On the bright side, the OSD stuff through the Tramp HV all works like it is supposed to.  

I've got a couple more XSRs on order, so I'll test them when they come in.

**Comment by Dutchy-79** [comment]
Then the firmware is right. 

I didn't touch halfduplex and left everything on stock settings. Just turned on telemetry in the configuration

**Comment by clcorbin** [comment]
Minor progress.  After resoldering the leads to the FC and shortening the leads by about half (I had originally soldered the full length XSR harness to it for testing, I've got a BIT better telemetry stability.  I was actually able to get the rates page of the LUA script to load several times as well as the VTR page.  Never anything on the PID page.

The telemetry was definitely coming in faster after shortening the header.  It STILL slows down and eventually stops sending FC telemetry.  I have four XSRs (bought a handful when they went on sale, just in case) and have tried three of them with the same results.  I have tried four different firmwares on two of them as well.  

I'm fairly sure it is a noise issue corrupting the signal.  I'm tempted to find the inverter on USART 1 of my Kakute F4 AIO FC and bypass it (and on the XSR) to see if that cleans things up a bit.  The idea of having a hardware inverter before transmission only to go into ANOTHER hardware inverter before going into the receiver (USART receiver, not quad copter XSR receiver!) is just stupid at best.

What a pain.  On the other hand, the OSD telemetry appears to be working...

**Comment by clcorbin** [comment]
After switching of softserial (LED pin) I was able to get it working 100% correctly.  Reading and writing data from all pages of the script work perfectly.  I decided to do a bit of testing with my scope.  The data coming in from off the softserial port was very consistent (values were changing, but timing and duration was consistent).

I then hooked things back up to the s.port on my Kakute F4 AIO and reconfigured Betaflight to use USART 1.  

The surprise was the what the scope showed.  The edges on the signal were great (no difference between softserial and USART1 through inverters).  What was REALLY different was how often data was being sent.  It was only sending data a tiny fraction of the time softserial was.  Sometimes, it wouldn't even get data drawn before it was erased by an empty frame.

Given the quality of the edges, I wouldn't think the issue is the inverters.  But I really wonder what is going on hardware wise in the F4 that is leading to all the missed data.  I wouldn't think the code is wrong as it works perfectly fine with so many other controllers.
